ssl/t1_lib.c

```diff
@@ -1239,117 +1217,97 @@ SSL_TICKET_RETURN tls_get_ticket_from_client(SSL *s, CLIENTHELLO_MSG *hello,
         return SSL_TICKET_NONE;
 
     size = PACKET_remaining(&ticketext->data);
-    if (size == 0) {
-        /*
-         * The client will accept a ticket but doesn't currently have
-         * one.
-         */
-        s->ext.ticket_expected = 1;
-        return SSL_TICKET_EMPTY;
-    }
-    if (s->ext.session_secret_cb) {
-        /*
-         * Indicate that the ticket couldn't be decrypted rather than
-         * generating the session from ticket now, trigger
-         * abbreviated handshake based on external mechanism to
-         * calculate the master secret later.
-         */
-        return SSL_TICKET_NO_DECRYPT;
-    }
 
-    retv = tls_decrypt_ticket(s, PACKET_data(&ticketext->data), size,
+    return tls_decrypt_ticket(s, PACKET_data(&ticketext->data), size,
                               hello->session_id, hello->session_id_len, ret);
-
-    /*
-     * If set, the decrypt_ticket_cb() is always called regardless of the
-     * return from tls_decrypt_ticket(). The callback is responsible for
-     * checking |retv| before it performs any action
-     */
-    if (s->session_ctx->decrypt_ticket_cb != NULL) {
-        size_t keyname_len = size;
-
-        if (keyname_len > TLSEXT_KEYNAME_LENGTH)
-            keyname_len = TLSEXT_KEYNAME_LENGTH;
-        retv = s->session_ctx->decrypt_ticket_cb(s, *ret,
-                                                 PACKET_data(&ticketext->data),
-                                                 keyname_len,
-                                                 retv, s->session_ctx->ticket_cb_data);
-    }
-
-    switch (retv) {
-    case SSL_TICKET_NO_DECRYPT:
-        s->ext.ticket_expected = 1;
-        return SSL_TICKET_NO_DECRYPT;
-
-    case SSL_TICKET_SUCCESS:
-        return SSL_TICKET_SUCCESS;
-
-    case SSL_TICKET_SUCCESS_RENEW:
-        s->ext.ticket_expected = 1;
-        return SSL_TICKET_SUCCESS;
-
-    case SSL_TICKET_EMPTY:
-        s->ext.ticket_expected = 1;
-        return SSL_TICKET_EMPTY;
-
-    case SSL_TICKET_NONE:
-        return SSL_TICKET_NONE;
-
-    default:
-        return SSL_TICKET_FATAL_ERR_OTHER;
-    }
 }
 
 /*-
  * tls_decrypt_ticket attempts to decrypt a session ticket.
  *
+ * If s->tls_session_secret_cb is set and we're not doing TLSv1.3 then we are
+ * expecting a pre-shared key ciphersuite, in which case we have no use for
+ * session tickets and one will never be decrypted, nor will
+ * s->ext.ticket_expected be set to 1.
```

Comment:
Why is this true?  Why can't we have a session_secret_cb and still be willing to do non-PSK handshakes?

Reply:
`session_secret_cb` is a bit odd. It is only there to support EAP-FAST. This has its own special implementation of pre-shared keys not based on the standard mechanism. If we're in an EAP-FAST setting then it only makes sense to do the EAP-FAST variant of TLS which implies using the EAP-FAST pre shared keys. It also implies special semantics for session tickets. Those tickets are not provisioned via the NewSessionTicket mechanism - the provisioning is done out-of-band. Incoming tickets, since they were not generated by OpenSSL in the first place, are not understood by us either. Therefore we do not decrypt them, nor do we set `ticket_expected` to 1 (which would have the impact of causing a NewSessionTicket to be sent).

Reply:
I'm about to sign-off on the PR, but I'd like to ask you to add a bit of commentary about this at these places, and perhaps above the definition of SSL_set_session_secret_cb().  Perhaps separately some day document this interface, unless it is intended to be only for users "in the know"...

Reply:
> I'd like to ask you to add a bit of commentary about this at these places, and perhaps above the definition of SSL_set_session_secret_cb().

Will do - but in a separate PR.